### PR TITLE
Guard closing quitCh with sync.Once to prevent double close

### PR DESCRIPTION
**Pull request by johejo**
related #8144 #8166 

See https://github.com/prometheus/prometheus/issues/8144#issuecomment-735814282

Signed-off-by: Mitsuo Heijo <mitsuo.heijo@gmail.com>

### CODE REVIEW

**Comment by roidelapluie** [comment]
You could instead revert 8166 and apply to the full function ?

pseudo code

```
var stopped bool
quitOnce.Do {
 stopped=true
  print(Quitting)
  close(quit)
}
if !stopped {
  print(Exit in progress)
}
```

**Comment by johejo** [comment]
Thanks good suggestion.
I think it's better only Handler.Quit(e.g. from main module) could read quitCh.

File: web/web.go [hunk 1]
```diff
@@ -918,12 +919,15 @@ func (h *Handler) version(w http.ResponseWriter, r *http.Request) {
...
 
 func (h *Handler) quit(w http.ResponseWriter, r *http.Request) {
-	select {
-	case <-h.quitCh:
+	var stopped bool
+	h.quitOnce.Do(func() {
+		stopped = true
+		close(h.quitCh)
+	})
+	if stopped {
```
Comment (roidelapluie) [inline web/web.go:927]: This should be the opposite

Comment (johejo) [reply web/web.go:927]: I squashed my fix, is it OK?

A
```
if cond {
} else {
}
```

B
```
if !cond {
} else {
}
```

If B is preferred over A in this project, I will fix further.

Comment (roidelapluie) [reply web/web.go:927]: Should we rename stopped to closed, and move the print inside the Do () ? Why do we have if else after where a simple if would be enough? Also is it safe to write after closing?

Comment (johejo) [reply web/web.go]: Fixed in https://github.com/prometheus/prometheus/pull/8242/commits/bc423903a5f59564215392576df82a48e09f2672

**Comment by roidelapluie** [comment]
Thanks!

**Pull request merged**